azure-cs-cosmosapp-component/VmScaleSets.cs

```diff
@@ -0,0 +1,293 @@
+﻿// Copyright 2016-2018, Pulumi Corporation.  All rights reserved.
+
+using System;
+using System.Collections.Generic;
+using System.Collections.Immutable;
+using System.IO;
+
+using Pulumi;
+using Pulumi.Azure.Compute;
+using Pulumi.Azure.Compute.Inputs;
+using Pulumi.Azure.Core;
+using Pulumi.Azure.Lb;
+using Pulumi.Azure.Lb.Inputs;
+using Pulumi.Azure.Monitoring;
+using Pulumi.Azure.Monitoring.Inputs;
+using Pulumi.Azure.Network;
+
+public static class VmScaleSets
+{
+    public static IDictionary<string, object> Run()
+    {
+        // Read a list of target locations from the config file:
+        // Expecting a comma-separated list, e.g., "westus,eastus,westeurope"
+        var locations = ImmutableList.Create(new Config().Require("locations").Split(","));
+
+        var resourceGroup = new ResourceGroup("cosmosvms-rg", new ResourceGroupArgs { Location = locations[0] });
+
+        var vmss = new CosmosApp("vmss",
+            new CosmosAppArgs
+            {
+                ResourceGroup = resourceGroup,
+                Locations = locations,
+                DatabaseName = "pricedb",
+                ContainerName = "prices",
+                Factory = new Builder(resourceGroup).BuildVMScaleSetApp,
+            });
+
+        return new Dictionary<string, object>
+        {
+            { "vmssEndpoint", Output.Format($"{vmss.Endpoint}/cosmos") }
+        };
+    }
+
+    private class Builder
+    {
+        private ResourceGroup resourceGroup;
+
+        public Builder(ResourceGroup resourceGroup)
+        {
+            this.resourceGroup = resourceGroup;
+        }
+
+        public Func<RegionalContext, IRegionalEndpoint> BuildVMScaleSetApp(GlobalContext context)
+        {
+            var options = (CustomResourceOptions)ResourceOptions.Merge(context.Options, new CustomResourceOptions { DeleteBeforeReplace = true });
```

Comment:
This line feels a bit too tricky: I have to pass `new CustomResourceOption` (not `new ResourceOptions`) and then cast the result to `CustomResourceOption`. Maybe it should be `= CustomResourceOption.Merge(...same thing...)` and not accept `ResourceOptions`? @CyrusNajmabadi

Reply:
I don't like that cast as well. Would prefer to have something cleaner and faster to read.